aoetools-36/aoeping.8
.TH aoeping 8
.SH NAME
aoeping \- simple communication with AoE device
.SH SYNOPSIS
.B aoeping [options] {shelf} {slot} {netif}
.fi
.SH DESCRIPTION
The
.IR aoeping (8)
program performs simple one or two-round-trip communication with an
ATA over Ethernet (AoE) device.  It creates and receives AoE packets
directly, using raw network sockets.
.PP
Running
.IR aoeping (8)
without command line arguments will result in a 
short usage summary being displayed.
.PP
The
.IR aoeping (8)
program will wait forever if it doesn't receive
an expected response.  The caller should use a time out to catch
this situation.
.SS Arguments
.TP
\fBshelf\fP
This should be the shelf address (major AoE address) of the AoE device
to communicate with.
.TP
\fBslot\fP
This should be the slot address (minor AoE address) of the AoE device
to communicate with.
.TP
\fBnetif\fP
The name of the ethernet network interface to use for AoE
communications, e.g., eth1.
.SS Options
.TP
\fB-i\fP
Issue an ATA "identify device" command after receiving the AoE
device's Config
Query response.  The "ident" response will be printed on standard
output as a hexadecimal dump.
.TP
\fB-I\fP
Issue an ATA "identify device" command after receiving the AoE
device's Config
Query response.  The "ident" response will be pretty-printed on standard
output as selected human-readable fields.
.TP
\fB-v\fP
Turn on 
more copious output, including a hexadecimal dump of the Config Query
response from the AoE device (see AoE spec at URL below).
.TP
\fB-s\fP
This option takes an argument.  The
argument is a decimal integer that specifies the number of seconds
that
.IR aoeping (8)
will wait for a response before timing out and
exiting with a non-zero status.
.TP
\fB-S\fP
This option takes an argument.  The
argument is the name of a SMART command to send to the disk.  The
SMART commands in the list below are supported.  If the command
requires data 
transfer, one sector (512 bytes) of data is always the amount
transferred.  If the command takes a parameter (for the Low LBA
register), then the name of the SMART command is immediately followed
by a colon and then a number, the value of the parameter, e.g., "-S
read_log:1".
.IP
  read_data
  offline_immediate
  read_log
  write_log
  enable
  disable
  return_status

For \fBwrite_log\fP,
.IR aoeping (8)
reads from
standard input the one sector of data to be
written to the specified log.

If the AoE device does not support SMART commands or if the command is
aborted, an error message
is printed to standard error and
.IR aoeping (8)
exits with a non-zero status.  A command may be aborted if SMART is
disabled on the device.

The
.IR aoeping (8)
command just sends and receives SMART commands, without
interpreting them.  See the ATA specification for more information on
using SMART.
.LP
.TP
\fB-t\fP
(This is an advanced feature.)  This option has an argument.  The
argument is a decimal integer that is used as the initial tag, with
the highest bit set, as 
the 
first tag in ATA commands.  Tags for subsequent ATA commands will be
incremented by one.
.TP
\fB-h\fP
Show a usage summary.
.SH EXAMPLE
In this example, the root user
uses
.IR aoeping (8)
to check for the presence of aoe device e10.9 on
network interface eth0.
.IP
.EX
.nf
bash# aoeping -v 10 9 eth0 | head
tag: 80000000
eth: eth0
shelf: 10
slot: 9
config query response:
00 0d 87 aa c9 00 00 10 04 00 11 1f 88 a2 18 00 
00 0a 09 01 00 00 00 00 00 03 30 08 00 10 00 04 
66 6f 6f 0a 00 ff ff ff ff ff ff ff ff ff ff ff 
ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff 
ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff 
.fi
.EE
.LP
The next example shows root making sure the disk on the e10.9 is still
responsive by issuing an ATA device identify command with a 20-second
timeout. 
.IP
.EX
.nf
bash# aoeping -i -s 20 \\
  10 9 eth0 > /dev/null \\
  && echo ok
ok
.fi
.EE
.LP
The next example uses SMART to determine whether the disk on e10.9
thinks it 
has exceeded its error threshold.  The ATA spec says that the LBA Mid
register will be 0x4f when the disk has not exceeded its error
threshold.
.IP
.EX
.nf
bash# aoeping -S return_status \\
  10 9 eth0 | grep 'LBA Mid: 0x4f' \\
  > /dev/null \\
  && echo ok
ok
.fi
.EE
.LP
Note that in a script, it would be prudent to specify and handle a
timeout.  Also, a good script would make sure the
\fBStatus\fP register does not have the error bit (bit zero) or the
device fault bit (bit 5) set.
.SH "SEE ALSO"
.IR aoe-discover (8),
.IR aoe-interfaces (8),
.IR aoe-mkdevs (8),
.IR aoe-mkshelf (8),
.IR aoe-stat (8),

\fIAoE (ATA over Ethernet)\fP: http://support.coraid.com/documents/AoEr10.txt,

\fIATA specification\fP
.SH AUTHOR
Ed L. Cashin (ecashin@coraid.com)


NEWS
-*- change-log -*-
2017-03-15 Ed Cashin <ed.cashin@acm.org>
	mothballing standalone driver
	version-v87moth

2017-02-22 Ed Cashin <ed.cashin@acm.org>
	merge v87pre3-bi_error for 4.4.x support
	version-87

2016-03-16 Ed Cashin <ed.cashin@acm.org>
	include "etherd/" in character device names for udev-systemd
	version-87pre3

2016-01-07 Ed Cashin <ed.cashin@acm.org>
	update driver version string in aoe.h
	version-87pre2

2015-12-13 Ed Cashin <ed.cashin@acm.org>
	correct conf/41 kernel version test (immutable biovec support)
	version-87pre1

2015-07-04 Ed Cashin <ed.cashin@acm.org>
	version-86

2015-04-13 Ed Cashin <ed.cashin@acm.org>
	add support for immutable biovecs with backward compatibility
	version-86pre1

2015-04-01 Ed Cashin <ed.cashin@acm.org>
	update documentation with support.coraid.com offline

2013-10-14 Sam Hopkins <sah@coraid.com>
	version-85

2013-09-17 Ed Cashin <ecashin@coraid.com>
	fix compat problem with building on Fedora 19

2013-09-17 Ed Cashin <ecashin@coraid.com>
	version-85pre1

2013-09-10 Ed Cashin <ecashin@coraid.com>
	use aoetools-36

2013-08-26 Ed Cashin <ecashin@coraid.com>
	update compatibility system
	drop support for kernels older than 2.6.18

2013-08-22 Ed Cashin <ecashin@coraid.com>
	get page head reference for compound page tails in biovecs
	remove warning-generating do-nothing term from example udev rules

2013-07-10 Ed Cashin <ecashin@coraid.com>
	add pkg subdirectory for distro-specific packaging files
	version-84

2013-07-02 Ed Cashin <ecashin@coraid.com>
	merge upstream patch: reservation of MAX_HEADER bytes in skbs
	version-84pre3

2013-06-19 Ed Cashin <ecashin@coraid.com>
	converge style (whitespace, etc.) towards kernel.org-distributed aoe
	version-84pre2

2013-06-04 Ed Cashin <ecashin@coraid.com>
	note in README: 2.6.16 will be oldest supported kernel
	version-84pre1

2013-05-26 Ed Cashin <ecashin@coraid.com>
	upstream change: use kmemdup when appropriate
	move code snippets to match order in kernel.org

2013-05-21 Ed Cashin <ecashin@coraid.com>
	do not use skb members after calling dev_queue_xmit
	update URLs in README

2013-02-11 Swanand Rao <swrao@coraid.com>
	fix bug where rmmod causes driver to panic
	version-83pre1
	version-83

2012-12-13 Ed Cashin <ecashin@coraid.com>
	fix hard-to-hit use-after-free bug found by Dan Carpenter

2012-10-30 Swanand Rao <swrao@coraid.com>
	add support for parallelization of I/O completion
	version-82pre2
	version-82

2012-09-20 Ed Cashin <ecashin@coraid.com>
	clean up kernel thread management routine
	use aoetools-35

2012-08-15 Ed Cashin <ecashin@coraid.com>
	revert get_page use; avoid using compound page tail for zero page
	fix aoe_ktio pausing of version 80
	version-81pre2
	version-81

2012-08-03 Ed Cashin <ecashin@coraid.com>
	use {get,put}_page for page _count, avoiding panic for compound pages
	version-81pre1

2012-07-17 Ed Cashin <ecashin@coraid.com>
	version-80

2012-06-29 Ed Cashin <ecashin@coraid.com>
	use aoetools-34
	cap I/O completion iterations before yeilding CPU in aoe_ktio
	support 3.2.y kernel
	version-80pre1

2012-02-09 Ed Cashin <ecashin@coraid.com>
	version-79

2012-01-27 Ed Cashin <ecashin@coraid.com>
	tell Linux virtual memory subsystem when aoe_tx writes
	version-79pre4

2011-12-30 Ed Cashin <ecashin@coraid.com>
	use aoetools-33
	update virtual memory settings information in HOWTO
	version-79pre3

2011-12-22 Ed Cashin <ecashin@coraid.com>
	avoid compiler warning in openSUSE 11.4
	update compat system: 1-param __blk_run_queue is latest
	version-79pre2

2011-12-14 Ed Cashin <ecashin@coraid.com>
	support openSUSE 11.4 in compatibility system
	version-79pre1

2011-12-08 Ed Cashin <ecashin@coraid.com>
	support 2.6.39 and newer kernels without big kernel lock
	version-78pre2
	version-78

2011-07-19 Ed Cashin <ecashin@coraid.com>
	support Scientific Linux 6.0 with check for build dependency: perl
	version-78pre1

2011-05-25 Ed Cashin <ecashin@coraid.com>
	support 2.6.38
	version-77

2010-11-19 Ed Cashin <ecashin@coraid.com>
	use aoetools-32
	avoid sending more than one probe at a time on tainted path
	get more I/O to do after stopping probes
	do not probe unless untainted aoetgts exist
	support 2.6.36
	version-76

2010-06-21 Ed Cashin <ecashin@coraid.com>
	support new name for blk_queue_max_sectors in 2.6.34
	get kmalloc and kfree declarations from slab.h to support 2.6.34
	version-75

2010-06-09 Ed Cashin <ecashin@coraid.com>
	avoid using system time in coarse-grained calculations
	support include/asm symlink removal in 2.6.33
	support SLES 10's unusual placement of skb_reset_mac_header

2010-01-08 Ed Cashin <ecashin@coraid.com>
	update copyright notices
	version-74

2009-11-23 Ed Cashin <ecashin@coraid.com>
	impose cap on retransmit time to single remote MAC address
	remove memory leak in grow_targets for more than 4 remote MACs
	ignore outstanding probes in decision to call I/O request handler
	only use one probe per tainted remote MAC address

2009-09-21 Ed Cashin <ecashin@coraid.com>
	do not run autoload.sh to configure loading of aoe on dev node access
	use aoetools-30
	support changes in 2.6.31
	version-73

2009-07-20 Ed Cashin <ecashin@coraid.com>
	use ATA_CMD_ values from ata.h instead of WIN_ macros in hdreg.h
	support older kernels where ata.h is incompatible with hdreg.h
	allow aoedbg.c to see DEFINE_SPINLOCK in pre 2.6.11 kernels
	version-72

2009-06-29 Ed Cashin <ecashin@coraid.com>
	do not use relay when Module.symvers is not present

2009-04-09 Ed Cashin <ecashin@coraid.com>
	avoid premature I/O failure: maintain timings on frame reassignment
	version-71

2009-02-25 Ed Cashin <ecashin@coraid.com>
	support commas instead of spaces in aoe_iflist parameter
	Roel Kluin: in skbfree, use skb if it's available on last iteration

2009-02-06 Ed Cashin <ecashin@coraid.com>
	fix the "kerninst" target for make, which had atrophied

2009-02-03 Ed Cashin <ecashin@coraid.com>
	fix and refine compatibility tests
	version-70

2009-01-28 Ed Cashin <ecashin@coraid.com>

aoetools-36/aoeping.c
/* aoeping.c - userland aoe pinger
 * Copyright 2009, CORAID, Inc., and licensed under GPL v.2.
 *
 * run without arguments for usage
 */
#include <stdio.h>
#include <string.h>
#include <stdlib.h>
#include <unistd.h>
#include <stdint.h>
#include <ctype.h>
#include <netinet/in.h>
#include <linux/hdreg.h>
#include <errno.h>
#include "dat.h"
#include "fns.h"

struct progopts {
	int shelf;
	int slot;
	char *netif;
	int verbose;
	int timeout;
	u32 tag;
	char *smart;
	char ata_ident;
	char pp_ataid;		/* pretty print ATA device identify response */
};

static struct progopts defaults = {
	.shelf = 0,
	.slot = 0,
	.netif = NULL,
	.verbose = 0,
	.timeout = 0,
	.tag = 0,
	.smart = NULL,
	.ata_ident = 0,
	.pp_ataid = 0,
};
static struct progopts opts;

struct smartcmd {
	char *name;		/* subcommand name from ATA spec */
	int cmd;		/* for features register */
	char data;		/* does this subcommand xfer data? */
};
static struct smartcmd smarts[] = {
	{ "read_data", 0xd0, SmartDataRet },
	// { "attr_autosave", 0xd2, 0 }, (unsupported b/c it overloads sector count)
	{ "offline_immediate", 0xd4, 0 },
	{ "read_log", 0xd5, SmartDataRet },
	{ "write_log", 0xd6, SmartDataPut },
	{ "enable", 0xd8, 0 },
	{ "disable", 0xd9, 0 },
	{ "return_status", 0xda, 0 },
};

static char *progname;
static int sfd;			/* raw socket file descriptor */
static uchar mac[6];

void
usage(void)
{
	fprintf(stderr,
		"usage:\t%s [options] {shelf} {slot} {netif}\n",
		progname);
	fprintf(stderr, "%s\n\t%s\n\t%s\n\t%s\n\t%s\n",
		"options:",
		"-i\tdo ATA device identify dump as raw hex",
		"-I\tdo ATA device identify print fields",
		"-v\tbe verbose",
		"-h\tshow this usage summary");
	fprintf(stderr, "%s\n\t%s\n\t%s\n\t%s\n",
		"options taking arguments:",
		"-s\ttimeout in seconds",
		"-S\tperform SMART command",
		"-t\tspecify number for starting AoE tag");
}

void
hex_print(FILE *out, uchar *buf, int n, char *sep)
{
	int i;
	int per_line = 16;

	for (i = 0; i < n;) {
		fprintf(out, "%02x%s", *buf++ & 0xff, sep);
		if (!(++i % per_line))
			putc('\n', out);
	}
}

void
find_blade(Conf *c, struct progopts *opts)
{
	int n;
	uchar buf[1400];
	u32 tag;

	Aoehdr *h = &c->h;

	memset(h, 0, sizeof *h);
	memset(h->dst, 0xff, sizeof h->dst);
	memmove(h->src, mac, sizeof h->src);
	
	h->type = htons(AOE_ETH_PROTO);
	h->flags = AoEver << 4;
	h->maj = htons(opts->shelf);
	h->min = opts->slot;
	h->cmd = Config;
	tag = htonl(opts->tag);
	memmove(h->tag, &tag, sizeof h->tag);
	c->bufcnt = 0;
	c->firmware = 0;
	c->scnt = 0;
	c->vercmd = Qread;
	c->len = htons(1024);
	memset(c->data, 0xED, sizeof c->data);
	if (write(sfd, c, sizeof *c) == -1) {
		perror("send config query");
		exit(EXIT_FAILURE);
	}
	for (;;) {
		n = read(sfd, buf, sizeof buf);
		if (n < 0) {
			perror("read network");
			exit(EXIT_FAILURE);
		}
		if (n < 60)
			continue;
		h = (Aoehdr *) buf;
		if (ntohs(h->type) != AOE_ETH_PROTO
			|| ntohs(h->maj) != opts->shelf
			|| h->min != opts->slot
			|| memcmp(h->tag, &tag, sizeof h->tag))
			continue;
		break;
	}
	if (opts->verbose) {
		puts("config query response:");
		hex_print(stdout, buf, n, " ");
		putchar('\n');
	}
	memcpy(c, buf, sizeof *c);
}

/* read a packet that was sent by the device that returned *c earlier
 */
int
aoe_pkt_read(uchar *buf, size_t siz, Conf *c, u32 tag)
{
	Aoehdr *h;
	int n;

	tag = htonl(tag);
	for (;;) {
		n = read(sfd, buf, siz);
		if (n < 0) {
			perror("read network");
			exit(EXIT_FAILURE);
		}
		if (n < 60)
			continue;
		h = (Aoehdr *) buf;
		if (ntohs(h->type) != AOE_ETH_PROTO
			|| h->maj != c->h.maj
			|| h->min != c->h.min
			|| memcmp(&tag, h->tag, sizeof h->tag))
			continue;
		break;
	}
	return n;
}

/* prepare a packet for doing ATA to a device that gave us Conf *c
 */
void
ata_prep(Ata *a, Conf *c, u32 tag)
{
	memset(a, 0, sizeof *a);
	memcpy(a->h.dst, c->h.src, sizeof a->h.dst);
	memcpy(a->h.src, mac, sizeof a->h.src);
	a->h.type = htons(AOE_ETH_PROTO);
	a->h.flags = AoEver << 4;
	a->h.maj = c->h.maj;
	a->h.min = c->h.min;
	a->h.cmd = ATAcmd;
	tag = htonl(tag);
	memmove(a->h.tag, &tag, sizeof a->h.tag);
}

/* pretty print ATA device identify text field
 * bytes have already been swapped
 */
void
pp_idtext(char *prefix, unsigned char *p, size_t len)
{
	int i;

	fputs(prefix, stdout);
	for (i = 0; i < len; ++i, ++p) {
		if (*p == '\0')
			break;
		if (!isgraph((int) *p) && *p != ' ')
			break;
		putchar(*p);
	}		
	putchar('\n');
}

int smart_supported(unsigned char *p)
{
	u16 w;

	p += 82 * 2;		/* skip to word 82 */
	w = *p++;
	w |= *p << 8;

	/* word 82 bit 0 is SMART support per ATA spec */
	return !!(w & 1);
}

void
disk_identify(Conf *c, struct progopts *opts, int *smart)
{
	int n;
	uchar buf[1400];
	Ata a;
	Ata *p;
	struct hd_driveid *id;

	ata_prep(&a, c, opts->tag);
	a.sectors = 1;
	a.cmd = ATAid_dev;
	a.lba[3] = 0xa0;

	if (write(sfd, &a, sizeof a) == -1) {
		perror("send ATA identify device");
		exit(EXIT_FAILURE);
	}

	n = aoe_pkt_read(buf, sizeof buf, c, opts->tag);
	p = (Ata *) buf;

	*smart = smart_supported(p->data);

	if (opts->ata_ident && !opts->pp_ataid) {
		puts("device identify response:");
		hex_print(stdout, p->data, 512, " ");
		return;
	}
	if (!opts->pp_ataid)
		return;
	
	for (n = 0; n < 1024; n += 2) {
		unsigned char ch;
		ch = p->data[n];
		p->data[n] = p->data[n+1];
		p->data[n+1] = ch;
	}
	id = (struct hd_driveid *) p->data;
	puts("device identify fields:");
	printf("vendor_specific_0: 0x%X\n", id->vendor0);
	printf("vendor_specific_1: 0x%X\n", id->vendor1);
	printf("vendor_specific_2: 0x%X\n", id->vendor2);
	pp_idtext("serial_number: ", id->serial_no, sizeof id->serial_no);
	pp_idtext("firmware_rev: ", id->fw_rev, sizeof id->fw_rev);
	pp_idtext("model: ", id->model, sizeof id->model);
}

struct smartcmd *
smartcmd_lookup(char *nam)
{
	int n = sizeof smarts / sizeof smarts[0];
	int i;

	for (i = 0; i < n; ++i) {
		char *p = strchr(nam, ':');

		if (p && !strncmp(smarts[i].name, nam, p - nam))
			return &smarts[i];
		else if (!strcmp(smarts[i].name, nam))
			return &smarts[i];
	}
	return nil;
}

void
smart_registers(Ata *a, char *opts, struct smartcmd *s)
{
	a->err = s->cmd;
	a->lba[1] = 0x4f;
	a->lba[2] = 0xc2;
	if (opts++)
		a->lba[0] = strtol(opts, NULL, 0);
}

int
show_smart_regs(Ata *a)
{
	if (a->err & ATAabrt) {
		fputs("SMART command aborted on target.\n",
		      stderr);
		return -1;
	}

	puts("ATA registers:");
	char *names[] = {
		"Features", "Sector Count",
		"LBA Low", "LBA Mid", "LBA High",
		"Status",
	};
	int regs[] = {
		a->err, a->sectors,
		a->lba[0], a->lba[1], a->lba[2],
		a->cmd,
	};
	int i;

	for (i = 0; i < sizeof regs / sizeof regs[0]; ++i)
		printf("%20s: 0x%02x\n", names[i], regs[i]);

	return 0;
}

void
smart(Conf *c, u32 tag, char *smart_cmd)
{
	int n;
	uchar buf[1400];
	Ata a;
	Ata *p;
	struct smartcmd *s = smartcmd_lookup(smart_cmd);

	if (!s) {
		fprintf(stderr,
			"%s Error: no such SMART command: %s\n",
			progname, smart_cmd);
		exit(EXIT_FAILURE);
	}

	ata_prep(&a, c, tag);
	a.sectors = !!s->data;	/* we only support one-sector data xfer */
	a.cmd = ATAsmart;
	smart_registers(&a, strchr(smart_cmd, ':'), s);
	if (s->data & SmartDataPut) {
		if (read(STDIN_FILENO, a.data, 512) == -1) {
			perror("reading smart data from stdin");
			exit(EXIT_FAILURE);
		}
		a.h.flags |= Write;
	}

	if (write(sfd, &a, sizeof a) == -1) {
		perror("send ATA identify device");
		exit(EXIT_FAILURE);
	}
	n = aoe_pkt_read(buf, sizeof buf, c, tag);
	p = (Ata *) buf;
	if (show_smart_regs(p) != 0)
		exit(EXIT_FAILURE);
	if (s->data & SmartDataRet) {
		puts("SMART data:");
		hex_print(stdout, p->data, 512, " ");
	}
}

void
bad_option(char c)
{
	fprintf(stderr, "%s Error: unrecognized option: ", progname);
	if (isprint(c))
		fprintf(stderr, "%c\n", c);
	else
		fprintf(stderr, "0x%02x\n", c & 0xff);
}

void
check_timeout(int secs)
{
	if (secs < 1) {
		fprintf(stderr,
			"%s Error: timeout seconds must be one or more\n",
			progname);
		exit(EXIT_FAILURE);
	}
}

void
init_opts(struct progopts *opts, int argc, char *argv[])
{
	int c;

	while ( (c = getopt(argc, argv, "hviIt:s:S:")) != -1) {
		switch (c) {
		case 'h':
			usage();
			exit(EXIT_SUCCESS);
			break;
		case 'v':
			++opts->verbose;
			break;
		case 'i':
			opts->ata_ident = 1;
			break;
		case 'I':
			opts->ata_ident = 1;
			opts->pp_ataid = 1;
			break;
		case 't':
			opts->tag = atoi(optarg);
			break;
		case 's':
			opts->timeout = atoi(optarg);
			check_timeout(opts->timeout);
			break;
		case 'S':
			opts->smart = optarg;
			break;
		case '?':
			bad_option(optopt);
			usage();
			exit(EXIT_FAILURE);
			break;
		default:
			abort(); /* shouldn't happen */
		}
	}
	if (argc - optind != 3) {
		usage();
		exit(EXIT_FAILURE);
	}
	opts->shelf = atoi(argv[optind]);
	opts->slot = atoi(argv[optind+1]);
	opts->netif = argv[optind+2];
}

int
main(int argc, char *argv[])
{
	Conf c;
	int smartable = 0;

	opts = defaults;
	progname = strrchr(argv[0], '/');
	if (progname)
		progname += 1;
	else
		progname = argv[0];

	init_opts(&opts, argc, argv);
	opts.tag |= 1UL << 31;	/* set high bit for userland AoE */

	if (opts.verbose) {
		printf("tag: %x\neth: %s\nshelf: %u\nslot: %u\n",
		       opts.tag, opts.netif, opts.shelf, opts.slot);
		fflush(stdout);
	}
	sfd = dial(opts.netif);
	if (!getea(sfd, opts.netif, mac))
		exit(EXIT_FAILURE);
	
	alarm(opts.timeout);
	find_blade(&c, &opts);
	opts.tag += 1;
	alarm(0);
	if (opts.verbose) {
		printf("found e%d.%d with mac ",
			ntohs(c.h.maj), c.h.min);
		hex_print(stdout, c.h.src, sizeof c.h.src, "");
		putchar('\n');
		fflush(stdout);
	}

	if (opts.ata_ident || opts.smart) {
		alarm(opts.timeout);
		disk_identify(&c, &opts, &smartable);
		alarm(0);
		opts.tag += 1;
	}

	if (opts.smart) {
		if (!smartable) {
			fprintf(stderr,
				"Error: e%d.%d does not support SMART\n",
				ntohs(c.h.maj), c.h.min);
			exit(EXIT_FAILURE);
		}
		alarm(opts.timeout);
		smart(&c, opts.tag, opts.smart);
		alarm(0);
		opts.tag += 1;
	}

	return 0;
}
